src/jdk.incubator.code/share/classes/jdk/incubator/code/internal/CodeModelToAST.java

```diff
@@ -0,0 +1,287 @@
+        Method method;
+        try {
+            method = invokeOp.invokeDescriptor().resolveToDirectMethod(MethodHandles.lookup());
+        } catch (ReflectiveOperationException e) {
+            throw new RuntimeException(e);
+        }
+        long flags = method.getModifiers();
+        var name = names.fromString(invokeOp.invokeDescriptor().name());
+        Value receiver = invokeOp.invokeKind() == CoreOp.InvokeOp.InvokeKind.INSTANCE ? invokeOp.operands().get(0) : null;
+        List<Value> arguments = invokeOp.operands().stream().skip(receiver == null ? 0 : 1).collect(List.collector());
+        var paramTypes = invokeOp.invokeDescriptor().type().parameterTypes().stream().map(this::typeElementToType)
+                .collect(List.collector());
+        var restype = typeElementToType(invokeOp.resultType());
+        var type = new Type.MethodType(paramTypes, restype, List.nil(), syms.methodClass);
+        var methodSym = new Symbol.MethodSymbol(flags, name, type,
+                typeElementToType(invokeOp.invokeDescriptor().refType()).tsym);
+        var meth = receiver == null ? treeMaker.Ident(methodSym) : treeMaker.Select((JCTree.JCExpression) valueToTree.get(receiver), methodSym);
+        var args = new ListBuffer<JCTree.JCExpression>();
+        for (Value operand : arguments) {
+            args.add((JCTree.JCExpression) valueToTree.get(operand));
+        }
+        var methodInvocation = treeMaker.App(meth, args.toList());
+        if (invokeOp.isVarArgs()) {
+            var lastParam = invokeOp.invokeDescriptor().type().parameterTypes().getLast();
+            Assert.check(lastParam instanceof ArrayType);
+            methodInvocation.varargsElement = typeElementToType(((ArrayType) lastParam).componentType());
+        }
+        if (invokeOp.result().uses().isEmpty()) {
+            return treeMaker.Exec(methodInvocation);
+        }
+        return methodInvocation;
+    }
+
+    private JCTree opToTree(Op op) {
+        JCTree tree = switch (op) {
+            case CoreOp.ConstantOp constantOp when constantOp.value() == null ->
+                    treeMaker.Literal(TypeTag.BOT, null).setType(syms.botType);
+            case CoreOp.ConstantOp constantOp -> treeMaker.Literal(constantOp.value());
+            case CoreOp.InvokeOp invokeOp -> invokeOpToJCMethodInvocation(invokeOp);
+            case CoreOp.NewOp newOp -> {
+                if (newOp.resultType() instanceof ArrayType at) {
+                    var elemType = treeMaker.Ident(typeElementToType(at.componentType()).tsym);
+                    // @@@ we assume one dimension
+                    var dims = List.of((JCTree.JCExpression) valueToTree.get(newOp.operands().getFirst()));
+                    var na = treeMaker.NewArray(elemType, dims, null);
+                    na.type = typeElementToType(at);
+                    yield na;
+                }
+                var ownerType = typeElementToType(newOp.constructorType().returnType());
```

Comment:
There is a strange asymmetry in the model between constructors and methods. Methods have a descriptor, constructors do not. Also, for methods you know from the op if it is varargs, but for constructors this info is not available. My feeling is that (at least coming at this from javac), trying to lump array and class creation under one op is probably why these asymmetries crop up.

This has nothing to do with this PR of course -- but I suspect that if you had a varargs constructor call this code would not work.

Reply:
I agree on all points. We can have a `new.array` operation and then model class creation as an `invoke` op ?

Reply:
> I agree on all points. We can have a `new.array` operation and then model class creation as an `invoke` op with method name being "init" ?

not sure I follow -- I was talking about class construction -- arrays are a bit special (e.g. there's no array constructor descriptor, unlike for classes). But a constructor call does share a lot of similarities with a method call, and there should be reusable code between method and constructors -- as it happens inside the javac code.

Reply:
This can be deferred until we have a better op to model instance creation expression.